### ISSUE TITLE
grub.cfg path incorrect when using BTRFS for root filesystem

### ISSUE THREAD

**Issue by luisgf** [OSInside/kiwi#1931]
## Problem description

Creating an ubuntu 20.04 image (may be applicable to other ubuntu versions) creates a wrong path in grub.cfg that leads to a image that doesn't boot properly when cloned to disk. 

Grub doesn't honor the default subvolume directive for a btrfs volume, so get_boot_path() should be take care of that and return a path that include the default volume @.  

The grub.cfg generated is this:

    set btrfs_relative_path="yes"
    search --fs-uuid --set=root ee00dc7e-b0c9-4b8e-ab7d-363af93f28b4
    set prefix=($root)/boot/grub
    configfile ($root)/boot/grub/grub.cfg

And the correct entry should be that:

    set btrfs_relative_path="yes"
    search --fs-uuid --set=root ee00dc7e-b0c9-4b8e-ab7d-363af93f28b4
    set prefix=($root)/@/boot/grub
    configfile ($root)/@/boot/grub/grub.cfg

## Expected behaviour
Image cloned to disk boots properly.

## Steps to reproduce the behaviour

Create a kiwi config.file with this configuration for storage:

    <preferences profiles="Disk">
          <type image="oem" filesystem="btrfs" firmware="uefi" installiso="true" btrfs_root_is_snapshot="false" installboot="install" installprovidefailsafe="false">
              <oemconfig>
                  <oem-swap>false</oem-swap>
                  <oem-boot-title>MyDevice</oem-boot-title>
                  <oem-unattended>false</oem-unattended>
                  <oem-swapsize>1024</oem-swapsize>
                  <oem-device-filter>/dev/ram</oem-device-filter>
                  <oem-multipath-scan>false</oem-multipath-scan>
              </oemconfig>
              <systemdisk name="vgsys">
                  <volume name="@root" size="10G"/>
                  <volume name="@home" mountpoint="/home" size="5G" />
                  <volume name="@tmp" mountpoint="/tmp" size="1G" />
                  <volume name="@varlog" mountpoint="/var/log" size="5G" copy_on_write="false" />
              </systemdisk>
          </type>
      </preferences>

## OS and Software information
+ KIWI version: 9.23.58
+ Operating system: Ubuntu 20.04
+ OBS version:

**Comment by luisgf** [comment]
Thanks @schaefi 

I have some patch to deal with this, but require some changes in get_boot_path(), passing one extra variable (with_subvol) that returns the complete volume path rather than the mountpoint. We can discuse that if you like.

**Comment by schaefi** [comment]
yeah would be super happy to see your patch :+1:

**Comment by schaefi** [comment]
I'm going to reproduce this today

**Comment by schaefi** [comment]
I can confirm the missing '@' in the early boot script. Interesting is that when I test this on Leap it boots in any case. Meaning

```
set btrfs_relative_path="yes"
search --fs-uuid --set=root 352ee2a1-58d4-4a6e-a21a-ea499c5430cf
set prefix=($root)/@/boot/grub2
configfile ($root)/@/boot/grub2/grub.cfg
```
behaves the same as

```
set btrfs_relative_path="yes"
search --fs-uuid --set=root 352ee2a1-58d4-4a6e-a21a-ea499c5430cf
set prefix=($root)/boot/grub2
configfile ($root)/boot/grub2/grub.cfg
```

which is different on ubuntu correct ?

**Comment by schaefi** [comment]
This makes me wonder under under which conditions a subvolume element '@' is relevant for grub ? The btrfs support in grub doesn't know about sub-volumes, at least this is what I thought. So it can only directly access data the "simple" way in the booting stage. For kiwi this was the reason why the early boot script just references the simple file path.

But I might be wrong in the way grub supports reading from btrfs. It would be good to really understand the background

**Comment by luisgf** [comment]
I think the key was in btrfs_relative_path, in ubuntu this option make nothins, perhaps in Suse can make some tricks with subvolumes.

**Comment by schaefi** [comment]
yep that's the case. Hrm, so this is a distro specific thing when it shouldn't be one right ? Still thinking how we can code this in a way that it does not harm on any side. I have tested with and without '@' in the path setting of the early boot script and it continues to work on suse. The btrfs_relative_path is only effective on suse I think and doesn't harm the others as it was there before. So my suggestion is we add the '@' for btrfs in a generic way. I assume your patch is doing this already ?

**Comment by schaefi** [comment]
an update on this one. I realized the tests I did were somewhat immaterial. The earlyboot script is only effective if kiwi needs to create a custom grub image. Before it does that it searches in the distribution for pre-build (package delivered) grub images which it will prefer. In case of my tests with Leap the leap provided grub image was used.

On ubuntu kiwi does not find a pre-build grub image (either because I'm searching at the wrong places or this does not exist) and therefore it creates its own grub image via grub-mkimage as you can see in the logs. In this case the earlyboot script becomes effective

I adapted my tests such that I can compare the situation for different distros

**Comment by schaefi** [comment]
I applied the '@' patch and this is what I get on leap

![grub](https://user-images.githubusercontent.com/912234/134360219-194b3bc0-fc93-463c-b6e9-1d820ced4ba2.png)

**Comment by schaefi** [comment]
not sure if this can be fixed without a distro switch :/

**Comment by schaefi** [comment]
although initially labeled as a bug, I don't think it is now. Looks more like a difference for which I currently have no good idea how to deal with in a generic way... hmm

**Comment by schaefi** [comment]
Still unsure how to move forward with this one and suffering from enoresponse ;)

**Comment by Conan-Kudo** [comment]
Reopening this because I got a question about this from @supakeen and did some spelunking.

The `btrfs_relative_path=` flag is used for boot-to-snapshot support, specifically snapper-style boot to snapshot. We probably should add a tunable for this because it only exists in distributions that implement SUSE-style boot to snapshot or carry the SUSE btrfs patches (like RH/Fedora GRUB does). It probably should be done alongside the work in #2697.

**Comment by schaefi** [comment]
From my perspective we addressed the mentioned issues. I'm closing this. We might want to add a btrfs based integration test for Ubuntu systems but I want to wait with this until their idea of the layout is set.
